tools/IoTEdgeDevOps/DevOpsLib/AzureActiveDirectory.cs

```diff
@@ -0,0 +1,112 @@
+// Copyright (c) Microsoft. All rights reserved.
+namespace DevOpsLib
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Net.Http;
+    using System.Net.Http.Headers;
+    using System.Threading.Tasks;
+    using Newtonsoft.Json.Linq;
+
+    public sealed class AzureActiveDirectory
+    {
+        static readonly AzureActiveDirectory instance = new AzureActiveDirectory();
+
+        static string azureActiveDirTenant = null;
+        static string azureActiveDirClientId = null;
+        static string azureActiveDirClientSecret = null;
+        static string azureResource = null;
+        static string accessToken = null;
+        static DateTime accessTokenExpiration = new DateTime(1970, 01, 01, 0, 0, 0, 0, DateTimeKind.Utc);
+
+        // Explicit static constructor to tell C# compiler
+        // not to mark type as beforefieldinit
+        static AzureActiveDirectory()
+        {
+        }
+
+        AzureActiveDirectory()
+        {
+        }
+
+        public static AzureActiveDirectory Instance
+        {
+            get
+            {
+                return instance;
+            }
+        }
+
+        // Trigger Azure Active Directory (AAD) for an OAuth2 client credential for an azure resource access.
+        // API reference: https://dev.loganalytics.io/documentation/Authorization/OAuth2
+        public async Task<string> GetAccessToken(
+            string azureActiveDirTenant,
+            string azureActiveDirClientId,
+            string azureActiveDirClientSecret,
+            string azureResource)
+        {
+            ValidationUtil.ThrowIfNullOrWhiteSpace(azureActiveDirTenant, nameof(azureActiveDirTenant));
+            ValidationUtil.ThrowIfNullOrWhiteSpace(azureActiveDirClientId, nameof(azureActiveDirClientId));
+            ValidationUtil.ThrowIfNullOrWhiteSpace(azureActiveDirClientSecret, nameof(azureActiveDirClientSecret));
+            ValidationUtil.ThrowIfNullOrWhiteSpace(azureResource, nameof(azureResource));
+
+            try
+            {
+                if ((DateTime.Compare(DateTime.UtcNow, AzureActiveDirectory.accessTokenExpiration) >= 0) ||
```

Comment:
you need to have lock statement to avoid race condition.

Reply:
I think you may need to add buffer in this checking? otherwise, access token may be expired right after this checking.  And the caller use this expired token, will get exception.